[FIRRTL] Update port symbols to InnerSymAttr

prithayan opened the pull request:
This PR updates ports to use `InnerSymAttr` for inner symbols instead of `StringAttr`.
`InnerSymAttr` makes it possible to add symbols per field.
This PR also updates the default symbol to `null` instead of an empty string name.
Before this PR, empty string was used to represent a port with no symbols, 
` portSyms = ["", "@sym1"]`
But this PR updates it to use `null` attribute instead:
` portSyms = [<<NULL ATTRIBUTE>>, [#firrtl<innerSym@sym1>]`
I am not sure if it is a good idea to add `null` attribute instead of the empty string in the IR.

dtzSiFive commented on lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp:
```diff
@@ -957,7 +960,7 @@ bool TypeLoweringVisitor::visitDecl(FExtModuleOp extModule) {
     newArgNames.push_back(port.name);
     newPortTypes.push_back(TypeAttr::get(port.type));
-    newArgSyms.push_back(port.sym ? port.sym : cache.sEmpty);
+    newArgSyms.push_back(port.sym ? port.sym : cache.symEmpty);
```
Is InnerSymAttr("") different than InnerSymAttr()?  And can this just be `port.sym`?

prithayan replied:
You are absolutely right, just using `port.sym`

dtzSiFive reviewed (commented):
Great!

Haven't finished going through, but generally looks good!  Will revisit soon!

Not sure about some code re:clearing out the port array, and generally treating "has symbol on fieldID=0" as "has worthwhile information to preserve", am I missing something?

By the way, do we want to verify these like we do for operation's?  IST's overall need to be verified, so maybe that is handled as part of that.

uenoku commented on include/circt/Dialect/FIRRTL/Namespace.h:
```diff
@@ -51,15 +51,16 @@ struct ModuleNamespace : public Namespace {
+    for (auto portSymbol : module.getPortSymbolsAttr())
+      if (portSymbol)
+        nextIndex.insert(
+            {portSymbol.cast<InnerSymAttr>().getSymName().getValue(), 0});
```
Agreed, I'm a bit concerned about `getSymName` because if it returns null for multiple per-field symbols, which might leads confusion in the future. (though we can fix later).

prithayan commented:
Thanks @dtzSiFive and @uenoku for the review.
I updated the PR to remove any empty string symbol names. Now the only check for a valid symbol is if the InnerSymAttr exists.
Also fixed the Namespace to iterate over all the field symbols for ports only.

prithayan commented on lib/Dialect/FIRRTL/Transforms/Dedup.cpp:
```diff
@@ -940,32 +940,36 @@ struct Deduper {
+      if (!fromPortSyms[portNo])
+        continue;
+      auto fromSym = fromPortSyms[portNo].cast<InnerSymAttr>();
+      if (!fromSym.hasValidSymName())
```
Removed the `hasValidSymname` completely.

prithayan commented on lib/Dialect/FIRRTL/FIRRTLAttributes.cpp:
```diff
@@ -122,7 +122,12 @@ StringAttr InnerSymAttr::getSymIfExists(unsigned fieldId) {
+
+bool InnerSymAttr::hasValidSymName() const {
+  auto s = getSymName();
+  return (s && !s.getValue().empty());
```
You are right, I removed the `hasValidSymname` completely and now just relying on the attribute only.

uenoku commented on lib/Conversion/FIRRTLToHW/LowerToHW.cpp:
```diff
@@ -897,7 +897,8 @@ LogicalResult FIRRTLModuleLowering::lowerPorts(
     hwPort.name = firrtlPort.name;
     hwPort.type = lowerType(firrtlPort.type);
-    hwPort.sym = firrtlPort.sym;
+    hwPort.sym = firrtlPort.sym ? firrtlPort.sym.getSymName()
+                                : StringAttr::get(moduleOp->getContext(), "");
```
Can you add a TODO comment or open an issue that field sensitive symbols are ignored at LowerToHW?

prithayan replied:
Will open the issue. Maybe we have to extend the field sensitive symbols for HW dialect also

prithayan replied:
Added an error here. I don't think it would be legal to ignore the field sensitive symbols here.

prithayan commented:
Updated the port symbols array to use empty string instead of null Attribute. 
This was done after @uenoku confirmed that MLIR convention is every element of an `ArrayAttr` must be non-null.

dtzSiFive commented on lib/Dialect/FIRRTL/FIRRTLOps.cpp:
```diff
@@ -205,6 +205,12 @@ void getAsmBlockArgumentNamesImpl(Operation *op, mlir::Region &region,
   }
 }
 
+static bool hasPortNamed(FModuleLike op, StringAttr name) {
```
This method is no longer needed (dropped with IST update).  FWIW it appears to also be the only user of `hasSymNamed` but that doesn't mean it's not useful.

prithayan replied:
Removed it.

dtzSiFive commented on lib/Dialect/FIRRTL/FIRRTLOps.cpp:
```diff
@@ -582,20 +579,13 @@ void FMemModuleOp::insertPorts(ArrayRef<std::pair<unsigned, PortInfo>> ports) {
   (*this)->setAttr("portTypes", ArrayAttr::get(getContext(), newTypes));
   (*this)->setAttr("portAnnotations", ArrayAttr::get(getContext(), newAnnos));
-  (*this)->setAttr("portSyms", ArrayAttr::get(getContext(), newSyms));
+  (*this).setPortSymbols(newSyms);
```
Nice! Love all these simplifications/cleanups!

dtzSiFive commented on include/circt/Dialect/FIRRTL/Namespace.h:
```diff
@@ -51,15 +51,17 @@ struct ModuleNamespace : public Namespace {
-      if (!portSymbol.getValue().empty())
-        nextIndex.insert({portSymbol.getValue(), 0});
+    for (auto portSymbol : module.getPortSymbolsAttr())
+      if (portSymbol)
```
No longer need to check if it's an empty string? Awesome, thanks!

dtzSiFive commented on lib/Dialect/FIRRTL/FIRRTLOps.cpp:
```diff
@@ -180,7 +180,7 @@ bool firrtl::hasDontTouch(Value value) {
   auto arg = value.dyn_cast<BlockArgument>();
   auto module = cast<FModuleOp>(arg.getOwner()->getParentOp());
-  return (!module.getPortSymbol(arg.getArgNumber()).empty());
+  return (module.hasPortSymbolAttr(arg.getArgNumber()));
```
```suggestion
  return module.hasPortSymbolAttr(arg.getArgNumber());
```

dtzSiFive commented on lib/Dialect/FIRRTL/InnerSymbolTable.cpp:
```diff
@@ -70,8 +70,8 @@ InnerSymbolTable::InnerSymbolTable(Operation *op) {
     // TODO: Add fields per port, once they work that way (use addSyms)
```
```suggestion
```
Thanks for updating this, please drop the corresponding TODO! :+1: .

dtzSiFive commented on lib/Dialect/FIRRTL/InnerSymbolTable.cpp:
```diff
@@ -118,9 +118,9 @@ StringAttr InnerSymbolTable::getInnerSymbol(InnerSymTarget target) {
+    if (!sym)
       return {};
-    return sym;
+    return sym.getSymName();
```
This needs to accomodate the field as well.  Previously the `isPort()` if/else set `base` (the `InnerSymAttr`) on both branches, and fell through to grabbing whatever field is targeted (as it presently done for non-ports), but simplified per style rules into current form.
If this description is unclear I can update this part if you'd like!

prithayan replied:
Thanks for explaining it. used the fieldID.

dtzSiFive commented on lib/Dialect/FIRRTL/InnerSymbolTable.cpp:
```diff
@@ -70,8 +70,8 @@ InnerSymbolTable::InnerSymbolTable(Operation *op) {
-        if (auto sym = p.value().sym; sym && !sym.getValue().empty())
-          addSym(p.value().sym, InnerSymTarget(p.index(), curOp));
+        if (auto sym = p.value().sym)
+          addSyms(sym, InnerSymTarget(p.index(), curOp));
```
Awesome, thanks!

dtzSiFive commented on lib/Dialect/FIRRTL/Transforms/Dedup.cpp:
```diff
@@ -940,36 +940,37 @@ struct Deduper {
 
       // Record the renaming.
-      renameMap[fromSym] = toSym;
+      renameMap[fromSym.getSymName()] = toSym.getSymName();
```
Does renameMap need to include per-field symbols now as well?

dtzSiFive commented on lib/Dialect/FIRRTL/Transforms/GrandCentral.cpp:
```diff
@@ -2225,8 +2211,10 @@ hw::InnerRefAttr GrandCentralPass::getInnerRefTo(Operation *op) {
+  return ::getInnerRefTo(module, portIdx, "",
+                         [&](FModuleLike mod) -> ModuleNamespace & {
+                           return getModuleNamespace(mod);
+                         });
```
:tada: nice cleanup!

dtzSiFive commented on lib/Dialect/FIRRTL/Transforms/InjectDUTHierarchy.cpp:
```diff
@@ -208,8 +208,8 @@ void InjectDUTHierarchy::runOnOperation() {
-    if (portSym && !portSym.getValue().empty())
-      dutPortSyms.insert(portSym);
+    if (portSym)
+      dutPortSyms.insert(portSym.getSymName());
```
(and this is okay to ignore the per-field symbols on ports?)

dtzSiFive commented on lib/Dialect/FIRRTL/Transforms/LowerTypes.cpp:
```diff
@@ -696,9 +696,8 @@ TypeLoweringVisitor::addArg(Operation *module, unsigned insertPt,
-  bool oldArgHadSym = oldArg.sym && !oldArg.sym.getValue().empty();
-  if (oldArgHadSym) {
-    symtmp = (oldArg.sym.getValue() + field.suffix).str();
+  if (oldArg.sym) {
```
(Will handling per-field symbols be needed here in the future?  There are outstanding PR's tackling lowering per-field symbols in lowering IIRC, just checking)

uenoku replied:
Once GrandCentral gets lowered during LowerAnnotation, we will see field sensitive symbols so it would be necessary in the near future.

dtzSiFive commented on lib/Dialect/FIRRTL/Transforms/ModuleInliner.cpp:
```diff
@@ -655,9 +655,11 @@ Inliner::mapPortsToWires(StringRef prefix, OpBuilder &b,
     StringAttr newSym;
-    StringAttr oldSym = portInfo[i].sym;
-    if (!oldSym.getValue().empty())
+    StringAttr oldSym;
```
Similarly, there's a TODO here to handle generating new names for field symbols on inline ports, right?

prithayan replied:
For all the passes we need to add support for per field symbols. Didn't want to tackle that in this PR. As that might require some update to logic for the different passes. This PR just assumes only `fieldID=0` symbol is relevant and ignores the rest. We need subsequent PRs to update all the passes to handle this.

dtzSiFive commented on test/Dialect/FIRRTL/errors.mlir:
```diff
@@ -119,7 +119,7 @@ firrtl.circuit "foo" {
 "firrtl.module"() ( {
   ^entry(%a: i1):
 }) {sym_name = "foo", portTypes = [!firrtl.uint], portDirections = 1 : i1,
-    portNames = ["in0"], portAnnotations = [], portSyms = [""]} : () -> ()
```
Nice!

dtzSiFive reviewed (changes requested):
Generally looks good! Please fix the `ArrayRef` to stack memory issue.

Added some comments/nits/suggestions elsewhere :+1:.

uenoku commented on include/circt/Dialect/FIRRTL/FIRRTLOpInterfaces.td:
```diff
@@ -252,6 +253,22 @@ def FModuleLike : OpInterface<"FModuleLike"> {
+    static void fixupPortSymsArray(SmallVectorImpl<Attribute> &syms,
+                                   MLIRContext * context) {
+      bool isValid = false;
+      for (size_t i=0, e = syms.size(); i < e ; ++i)
```
nit: format
```suggestion
      for (size_t i = 0, e = syms.size(); i < e; ++i)
```

uenoku commented on lib/Conversion/FIRRTLToHW/LowerToHW.cpp:
```diff
@@ -897,7 +897,8 @@ LogicalResult FIRRTLModuleLowering::lowerPorts(
     hwPort.name = firrtlPort.name;
     hwPort.type = lowerType(firrtlPort.type);
-    hwPort.sym = firrtlPort.sym;
+    hwPort.sym = firrtlPort.sym ? firrtlPort.sym.getSymName()
+                                : StringAttr::get(moduleOp->getContext(), "");
```
Could you emit warnings if firrtlPort has field sensitive symbols? (since we are currently dropping field sensitive symbols at LowerToHW)

prithayan replied:
Yes, that and the I need to create an issue. Lets schedule a meeting this week and discuss the path forward here.

uenoku replied:
Thank you for emitting errors!

uenoku commented on include/circt/Dialect/FIRRTL/FIRRTLOpInterfaces.td:
```diff
@@ -161,36 +169,29 @@ def FModuleLike : OpInterface<"FModuleLike"> {
-        return StringAttr::get($_op.getContext(), "");
-      return syms[portIndex].template cast<StringAttr>();
+      if (syms.empty() ||
+          syms[portIndex].template cast<InnerSymAttr>().isSymbolInvalid())
```
Can we reject invalid symbols at the verifier level?

prithayan replied:
Unfortunately, invalid just means symbol name is empty string. And as we discussed we need empty string symbol names because we can't have null attributes. So, we need to add `""` as symbol name and then use `isSymbolInvalid` to ignore them.

dtzSiFive commented:
FWIW I don't think anything invokes the InnerSymAttr verifier (`verifyInnerSymAttr()`) on these Port InnerSymAttr's yet?  I guess it might belong as part of the IST verification (`InnerSymbolTable::verifyRegionTrait`), since only the IST can have port symbols (don't love hardcoding this konwledge everywhere, but it's what it is).  WDYT?

Duplicate symbols within and across these will be checked via #3609, but not any of the other checks in the verifier.

uenoku commented:
> FWIW I don't think anything invokes the InnerSymAttr verifier (verifyInnerSymAttr()) on these Port InnerSymAttr's yet? I guess it might belong as part of the IST verification (InnerSymbolTable::verifyRegionTrait), since only the IST can have port symbols (don't love hardcoding this konwledge everywhere, but it's what it is). WDYT?


Ah, thanks! I thought we already have had verifier enabled. If it is not enabled yet, I think it makes sense to have `isSymbolInvalid` for now.

prithayan commented on lib/Dialect/FIRRTL/FIRRTLAttributes.cpp:
```diff
@@ -131,6 +145,14 @@ bool InnerSymAttr::all_of_props(
 
+bool InnerSymAttr::isSymbolInvalid() {
+  return all_of_props([](InnerSymPropertiesAttr p) {
+    if (p.getName().getValue().empty())
```
Thanks fixed it.

prithayan commented on lib/Dialect/FIRRTL/FIRRTLAttributes.cpp:
```diff
@@ -122,7 +122,21 @@ StringAttr InnerSymAttr::getSymIfExists(unsigned fieldId) {
+  SmallVector<StringAttr> symNames;
+  for (auto p : getImpl()->props)
+    symNames.push_back(p.getName());
+  return llvm::makeArrayRef(symNames);
```
Thanks for catching this, just returning the vector copy instead.

prithayan commented on include/circt/Dialect/FIRRTL/FIRRTLAttributes.td:
```diff
@@ -232,20 +232,28 @@ def InnerSymAttr : AttrDef<FIRRTLDialect, "InnerSym"> {
+    bool all_of_props(std::function<bool (InnerSymPropertiesAttr)>);
     /// Get the number of inner symbols defined.
-    size_t numSymbols() ;
+    size_t numSymbols();
```
Done

prithayan commented on include/circt/Dialect/FIRRTL/FIRRTLOpInterfaces.h:
```diff
@@ -61,6 +61,28 @@ struct PortInfo {
+  /// Default constructors
+  PortInfo(StringAttr name, FIRRTLType type, Direction dir)
+      : name(name), type(type), direction(dir) {}
+  PortInfo(StringAttr name, FIRRTLType type, Direction dir, StringAttr symName)
```
Thanks for the suggestion, combined all the redundant constructors.

prithayan commented:
Thanks @uenoku  and @dtzSiFive for yet another round of detailed review. Hopefully this is ready for the next round of reviews.
Actually `verifyModuleLikeOpInterface(FModuleLike module)` does verify that the ports are non null `InnerSymAttr` `attribute`, I don't think it calls the `verifyInnerSymAttr`. But, a very trivial verification is already there. Yes, the IST verification does get triggered, so that should be good.
But we can't check for `isSymbolInvalid` , because we have to store empty string as symbol names instead of null attribute.

There's also a whole bunch of followup work pending to fixup all the remaining passes to handle the field sensitive symbols. This PR just assumes only `fieldID=0` symbol is interesting, and ignore everything else.

uenoku reviewed (commented):
Thank you for updating this! It would be necessary to rebase PR regarding dontTouch representation changes.

uenoku commented on lib/Dialect/FIRRTL/FIRRTLOpInterfaces.cpp:
```diff
@@ -81,9 +81,10 @@ LogicalResult circt::firrtl::verifyModuleLikeOpInterface(FModuleLike module) {
-  if (llvm::any_of(portSymbols.getValue(),
-                   [](Attribute attr) { return !attr.isa<StringAttr>(); }))
-    return module.emitOpError("port symbols should all be string attributes");
+  if (llvm::any_of(portSymbols.getValue(), [](Attribute attr) {
+        return !attr || !attr.isa<InnerSymAttr>();
+      }))
```
Maybe it would be nice to check `attr.cast<InnerSymAttr>().isValid`(`!attr.cast<InnerSymAttr>().empty()`).

prithayan replied:
Actually, we have empty or invalid entries in the `portSymbols` now. It is required to store ports with no symbol in the `portSymbols` array.

uenoku commented on include/circt/Dialect/FIRRTL/InnerSymbolTable.h:
```diff
@@ -182,6 +182,30 @@ class InnerSymbolTable {
   /// A successful walk with no failures returns success.
   static LogicalResult walkSymbols(Operation *op, InnerSymCallbackFn callback);
 
+  /// Get the sym name from the InnerSymPropertiesAttr.
+  static StringAttr getInnerSymName(const InnerSymPropertiesAttr p) {
+    return p.getName();
+  }
+
+  /// Get the iterator for all the symbol names for an InnerSymAttr.
+  static auto getSymNamesBegin(InnerSymAttr isa) {
+    return llvm::map_iterator(isa.begin(), getInnerSymName);
+  }
+
+  /// Get the iterator for all the symbol names for an InnerSymAttr.
+  static auto getSymNamesEnd(InnerSymAttr isa) {
+    return llvm::map_iterator(isa.end(), getInnerSymName);
+  }
+
+  /// Get all the symbol names for an Operation.
+  static auto getSymNames(Operation *op) {
+    InnerSymAttr isa =
+        op->getAttrOfType<InnerSymAttr>(getInnerSymbolAttrName());
+    if (isa)
+      llvm::make_range(getSymNamesBegin(isa), getSymNamesEnd(isa));
+    return StringAttr{};
+  }
+
```
These can be removed now, right?

prithayan replied:
Thanks, I missed it. Removed them

dtzSiFive commented on lib/Dialect/FIRRTL/FIRRTLOps.cpp:
```diff
@@ -179,7 +179,7 @@ bool firrtl::hasDontTouch(Value value) {
   auto arg = value.dyn_cast<BlockArgument>();
   auto module = cast<FModuleOp>(arg.getOwner()->getParentOp());
-  return (!module.getPortSymbol(arg.getArgNumber()).empty()) ||
+  return (module.getPortSymbolAttr(arg.getArgNumber())) ||
```
Could this be `hasPortSymbolAttr`?
